File: 20162-INE5424-5-4-2-EPOS-Android-BD/Software/SOServer/src/main/java/br/com/soserver/serialization/Serialization.java
package br.com.soserver.serialization;
import org.apache.commons.lang3.StringUtils;

import java.lang.NumberFormatException;

public class Serialization {

	public String serialize(Message message) {
		String data = message.getData();
		int ldata = data.length();
		String tmp = message.getAddress() + message.getFunctionCode() + data;
		int checksum = calculateLRC(tmp);
		String result = ":" + message.getAddress() + message.getFunctionCode()
										+ data + StringUtils.leftPad(Message.toHex(checksum), 2, '0')+ "\r\n";
		return result;
	}
	public Message deserialize(String content) {
		Message result = new Message();
		result.setAddress(content.substring(1,3));
		result.setFunctionCode(content.substring(3,5));
		int ind = content.lastIndexOf('\r');
		if (ind == -1) {
			return null;
		}

		result.setDataSlice(content.substring(5, ind-2), 0, ind-7);
		int checksum;
		try {
			checksum = result.fromHex(content.substring(ind-2, ind));
		} catch (NumberFormatException e) {
			return null;
		}

		String tmp = result.getAddress() + result.getFunctionCode() + result.getData();
		int checksum2 = calculateLRC(tmp);
		if (checksum2 != checksum || checksum2 == -1) {
			result = null;
		}
		return result;
	}
	public static int calculateLRC(String b) {
		try {
			int LRC = 0;
			int l = b.length();
			for (int i = 0; i < l; i += 2) {
				LRC += Message.fromHex(b.substring(i, i+2));
			}
			return ((byte) LRC) & 0xff;
		}catch (NumberFormatException e){
			return 0;
		}
	}
		// unsigned char nLRC = 0; // LRC char initialized
		// int l = strlen(data);
		// for (int i = 0; i < l; i += 2) {
		//   char *s = Message::sliceChar(data, i, i+2);
		//   nLRC -= Message::fromHex(s);
		//}
		//delete [] data;
}

File: 20162-INE5424-5-4-2-EPOS-Android-BD/Software/epos/serialization/serialization.cc
#ifndef SERIALIZATION_CC
#define SERIALIZATION_CC

#include "serialization.h"
#include "message.h"
#ifdef SIMULATE
  #include <cstring>
#else
  #include <utility/string.h>
#endif
const char* Serialization::serialize(Message *message) {
  const char* data = message->getData();
  int ldata = strlen(data);
  char* tmp = new char[5+ldata];
  memset(tmp, '\0', 5+ldata);
  strcat(tmp, message->getAddress());
  strcat(tmp, message->getFunctionCode());
  strcat(tmp, data);
  int checksum = this->lrc(tmp);
  delete[] tmp;
  char *result = new char[10+ldata];
  memset(result, '\0', 10+ldata);
  strcat(result, ":");
  strcat(result, message->getAddress());
  strcat(result, message->getFunctionCode());
  strcat(result, data);
  const char *checksumHex = Message::prefixPadding('0', Message::toHex(checksum), 2);
  strcat(result, checksumHex);
  strcat(result, "\r\n");
  delete[] checksumHex;
  return result;
}

Message* Serialization::deserialize(const char *content) {
  Message* result = new Message();
  result->setAddress(Message::sliceChar(content, 1, 3));
  result->setFunctionCode(Message::sliceChar(content, 3, 5));
  const char *r = strrchr(content, '\r');
  if (!r) {
    delete result;
    result = 0;
    return result;
  }
  int t = r-content;
  result->setDataSlice(Message::sliceChar(content, 5, t-2), 0, t-7);
  const char* checksumContent = Message::sliceChar(content, t-2, t);
  int checksum = Message::fromHex(checksumContent);
  delete[] checksumContent;
  // Check cheksum..
  char *tmp = new char[4+(t-7)+1];
  memset(tmp, '\0', 4+(t-7)+1);
  strcat(tmp, result->getAddress());
  strcat(tmp, result->getFunctionCode());
  strcat(tmp, result->getData());
  int checksum2 = this->lrc(tmp);
  delete[] tmp;
  if (checksum != checksum2) {
    delete result;
    result = 0;
  }
  return result;
}

int Serialization::lrc(char *data)
{
  int nLRC = 0; // LRC char initialized
  int l = strlen(data);
  for (int i = 0; i < l; i += 2) {
    const char *s = Message::sliceChar(data, i, i+2);
    nLRC += Message::fromHex(s);
    delete[] s;
  }
  return nLRC & 0xff;
}

#endif
